File: acwing/_1275.cpp
#include <cstdio>
#include <algorithm>

using namespace std;

const int N = 2e5 + 10;
int m, p, a, n;
struct Node {
    int l, r, mx;
} tr[N * 4];

void pushup(int u) {
    tr[u].mx = max(tr[u << 1].mx, tr[u << 1 | 1].mx);
}

void build(int u, int l, int r) {
    tr[u].l = l, tr[u].r = r;
    if (l == r) return; // 这题数组中没初值 找到叶子节点了可以不赋值
    int m = l + r >> 1;
    build(u << 1, l, m);
    build(u << 1 | 1, m + 1, r);
    // pushup(u); // 同上可以不做pushup操作
}

int query(int u, int l, int r) {
    if (tr[u].l >= l && tr[u].r <= r) return tr[u].mx;
    int m = tr[u].l + tr[u].r >> 1;
    int res = 0;
    if (l <= m) res = max(res, query(u << 1, l, r));
    if (r >= m + 1) res = max(res, query(u << 1 | 1, l, r));
    return res;
}

void modify(int u, int i, int x) {
    if (tr[u].l == tr[u].r) tr[u].mx = x;
    else {
        int m = tr[u].l + tr[u].r >> 1;
        if (i <= m) modify(u << 1, i, x);
        else modify(u << 1 | 1, i, x);
        pushup(u);
    }
}

int main() {
    scanf("%d%d", &m, &p);
    char op[2];
    int x;
    build(1, 1, m);
    while (m-- > 0) {
        scanf("%s%d", op, &x);
        if (op[0] == 'Q') {
            a = query(1, n - x + 1, n);
            printf("%d\n", a);
        } else modify(1, ++n, (1LL * x + a) % p);
        // 由于输入的数x和上一次查询到的数a范围都是[1,2e9] 这里查到的数可能会爆int
    }

    return 0;
}

File: acwing/_1275.java
import java.util.*;
import java.io.*;

public class _1275 {
    static int m, p, a, n;
    static class Node {
        int l, r, mx;

        Node(int l, int r) {
            this.l = l;
            this.r = r;
        }
    }
    static Node[] tr;

    static void pushup(int u) {
        tr[u].mx = Math.max(tr[u << 1].mx, tr[u << 1 | 1].mx);
    }

    static void build(int u, int l, int r) {
        tr[u] = new Node(l, r);
        if (l == r) return;
        int m = l + r >> 1;
        build(u << 1, l, m);
        build(u << 1 | 1, m + 1, r);
    }

    static int query(int u, int l, int r) {
        if (tr[u].l >= l && tr[u].r <= r) return tr[u].mx;
        int m = tr[u].l + tr[u].r >> 1;
        int res = 0;
        if (l <= m) res = query(u << 1, l, r);
        if (r >= m + 1) res = Math.max(res, query(u << 1 | 1, l, r));
        return res;
    }

    static void modify(int u, int i, int x) {
        if (tr[u].l == tr[u].r) tr[u].mx = x;
        else {
            int m = tr[u].l + tr[u].r >> 1;
            if (i <= m) modify(u << 1, i, x);
            else modify(u << 1 | 1, i, x);
            pushup(u);
        }
    }

    public static void main(String[] args) throws IOException {
        StreamTokenizer in = new StreamTokenizer(new BufferedReader(new InputStreamReader(System.in)));
        in.nextToken(); m = (int) in.nval;
        in.nextToken(); p = (int) in.nval;
        tr = new Node[m * 4];
        build(1, 1, m);
        while (m-- > 0) {
            in.nextToken(); char op = in.sval.charAt(0);
            in.nextToken(); int x = (int) in.nval;
            if (op == 'Q') {
                a = query(1, n - x + 1, n);
                System.out.println(a);
            } else modify(1, ++n, (int) ((1L * x + a) % p));
        }
    }
}
